plugin/evm/syncervm_client.go

```diff
@@ -317,6 +333,130 @@ func (client *stateSyncerClient) syncStateTrie(ctx context.Context) error {
+	if !client.blockBackfillEnabled {
+		return ids.Empty, 0, block.ErrBlockBackfillingNotEnabled
+	}
+
+	var (
+		nextBlkID     ids.ID
+		nextBlkHeight uint64
+	)
+	switch lastBackfilledBlkID, err := client.metadataDB.Get(lastBackfilledBlockKey); err {
+	case database.ErrNotFound:
+		// backfill was never attempted. Start from state sync block
+		var (
+			summaryBlkID     = client.resumableSummary.BlockHash
+			summaryBlkHeight = client.resumableSummary.BlockNumber
+		)
+		summaryBlk := rawdb.ReadBlock(client.chaindb, summaryBlkID, summaryBlkHeight)
+		if summaryBlk != nil {
+			return ids.Empty, 0, fmt.Errorf("failed retrieving summary block %s", summaryBlkID)
+		}
+
+		client.latestBackfilledBlock = ids.ID(summaryBlk.Hash())
+		nextBlkID = ids.ID(summaryBlk.ParentHash())
+		nextBlkHeight = summaryBlk.NumberU64() - 1
+
+	case nil:
+		// backfill was attempted. Resume from latest backfilled block
+		latestBackfilledBlock, err := client.getBlk(ctx, ids.ID(lastBackfilledBlkID))
+		if err != nil {
+			return ids.Empty, 0, fmt.Errorf("failed retrieving latest backfilled block, %s, %w", client.latestBackfilledBlock, err)
+		}
+
+		if latestBackfilledBlock.Height() == 1 {
+			return ids.Empty, 0, block.ErrStopBlockBackfilling
+		}
+
+		client.latestBackfilledBlock = ids.ID(lastBackfilledBlkID)
+		nextBlkID = latestBackfilledBlock.Parent()
+		nextBlkHeight = latestBackfilledBlock.Height() - 1
+
+	default:
+		return ids.Empty, 0, fmt.Errorf("failed retrieving last backfilled block ID from disk: %w", err)
+	}
+
+	log.Info(
+		"Starting block backfilling",
+		"next requested block ID", nextBlkID,
+		"next requested block height", nextBlkHeight,
+	)
+	return nextBlkID, nextBlkHeight, nil
+}
+
+func (client *stateSyncerClient) BackfillBlocks(ctx context.Context, blksBytes [][]byte) (ids.ID, uint64, error) {
+	// 1. Parse blocks
+	blks := make(map[uint64]snowman.Block)
+	for i, blkBytes := range blksBytes {
+		blk, err := client.parseBlk(ctx, blkBytes)
+		if err != nil {
+			log.Warn(
+				"Failed parsing backfilled block",
+				"block index", i,
+			)
+			continue
+		}
+		blks[blk.Height()] = blk
+	}
+
+	// 2. Validate blocks continuity and store them
+	blkHeights := maps.Keys(blks)
+	sort.Slice(blkHeights, func(i, j int) bool {
+		return blkHeights[i] < blkHeights[j] // sort in ascending order by heights
+	})
+
+	topBlk, err := client.getBlk(ctx, client.latestBackfilledBlock)
+	if err != nil {
+		return ids.Empty, 0, fmt.Errorf("failed retrieving latest backfilled block, %s, %w", client.latestBackfilledBlock, err)
+	}
+	for i := len(blkHeights) - 1; i >= 0; i-- {
+		blk := blks[blkHeights[i]]
+		if topBlk.Parent() != blk.ID() {
+			log.Warn(
+				"Unexpected backfilled block. Reissuing request",
+				"expected block ID", topBlk.Parent(),
+				"received block ID", blk.ID(),
+				"Requested block ID", topBlk.Parent(),
+			)
+			break
+		}
+
+		if err := blk.Accept(ctx); err != nil {
```

Comment:
This is most likely wrong. Accepting a past block seems to mess up a whole bunch of metadata.
@darioush  I'd love a feedback on how the indexing of past block should be handled (brand new methods? Upgrade block.Accept to handle past blocks? something else?).